### PR TITLE
Prepare for pypi package publishing

### PR DESCRIPTION
**Pull request by Liam-DeVoe**
this is effectively a rewrite of the codebase (not that there was much to rewrite to begin with).

I might suggest we think of a better name for this repo - "planetpy" perhaps? "PlanetTerp API Python Wrapper" is a bit of a mouthful. I think we should still name our pypi package "planetterp" regardless.

### CODE REVIEW

File: planetterp/PlanetTerp.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    """
+    Get all courses meeting some criteria.
+
+    Parameters
+    ----------
+    department: string
+        Only return courses in the given department. The department name must be
+        four characters. Defaults to all departments.
+    reviews: bool
+        Also return reviews for the course, specifically reviews for
```
Comment (saewitz) [inline planetterp/PlanetTerp.py:36]: "Also" -> "Whether to also" (consistent with course endpoint)

File: planetterp/PlanetTerp.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        Whether to also return reviews for the course, specifically reviews for
+        professors that taught the course and have the course listed as the one
+        being reviewed.
+    """
+    params = {"name" : name, "reviews": "true" if reviews else "false"}
+    url = BASE_URL + "course?" + urlencode(params)
+    return requests.get(url).json()
+
+
+def all_courses(department = None, reviews = False, limit = 100, offset = 0):
```
Comment (saewitz) [inline planetterp/PlanetTerp.py]: `all_courses` -> `courses` (that's the name in the api)

File: planetterp/PlanetTerp.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    """
+    Get all courses meeting some criteria.
+
+    Parameters
```
Comment (saewitz) [inline planetterp/PlanetTerp.py:30]: all 4 parameters should be optional

File: planetterp/PlanetTerp.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    ----------
+    department: string
+        Only return courses in the given department. The department name must be
+        four characters. Defaults to all departments.
+    reviews: bool
+        Also return reviews for the course, specifically reviews for
+        professors that taught the course and have the course listed as the one
+        being reviewed.
```
Comment (saewitz) [inline planetterp/PlanetTerp.py]: Defaults to false.

File: planetterp/PlanetTerp.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    reviews: bool, optional
+        Whether to also return reviews for the given professor. Defaults to
+        false.
+    """
+    params = {"name" : name, "reviews": "true" if reviews else "false"}
+    url = BASE_URL + "professor?" + urlencode(params)
+    return requests.get(url).json()
+
+
+def all_professors(type_ = None, reviews = False, limit = None, offset = None):
```
Comment (saewitz) [inline planetterp/PlanetTerp.py]: `all_professors` -> `professors` (that's the name in the api)

File: planetterp/PlanetTerp.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    """
+    Get all professors meeting some criteria.
+
+    Parameters
```
Comment (saewitz) [inline planetterp/PlanetTerp.py:78]: all 4 parameters should be optional

File: planetterp/PlanetTerp.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+        inclusive. Defaults to 100.
+    offset: int
+        Number of courses to skip (offered for pagination). Defaults to 0.
+    """
+    if type_ not in ["professor", "ta"]:
+        raise ValueError("Expected type to be on of ['professor', 'ta'], got "
+                         + str(type_))
+    params = {"type" : type_, "reviews": "true" if reviews else "false",
+              "limit": limit, "offset": offset}
+    # fitler out None args
```
Comment (saewitz) [inline planetterp/PlanetTerp.py]: why is this needed?

File: planetterp/PlanetTerp.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+    type_: {"professor", "ta"}
+        Only return reviews for the specified instructor type, either professors
+        or TAs. Defaults to returning both.
+    reviews: bool
+        Also return reviews for the instructors. Defaults to false.
+    limit: int
+        Maximum number of courses to return. Must be between 1 and 1000
+        inclusive. Defaults to 100.
+    offset: int
+        Number of courses to skip (offered for pagination). Defaults to 0.
```
Comment (saewitz) [inline planetterp/PlanetTerp.py]: "Number of courses" -> "Number of professors"

File: planetterp/PlanetTerp.py [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+    url = BASE_URL + "professors?" + urlencode(params)
+    return requests.get(url).json()
+
+
+def grades(course = None, professor = None, semester = None, section = None):
+    """
+    Get grades for a given course, professor, semester, section, or combination
+    thereof.
+
+    Parameters
```
Comment (saewitz) [inline planetterp/PlanetTerp.py:110]: Missing section parameter

Comment (saewitz) [inline planetterp/PlanetTerp.py:110]: Should clarify that one of `course` and `professor` is required. `semester` and `section` should be optional.

File: setup.py [hunk 10]
```diff
@@ -0,0 +1,25 @@
+import setuptools
+
+with open("README.md", "r", encoding="utf-8") as f:
+    long_description = f.read()
+
+__version__ = "1.0.0"
```
Comment (saewitz) [inline setup.py]: why is version defined in \_\_init__.py and in setup.py?

File: planetterp/PlanetTerp.py (hunk 7 above)
Comment (Liam-DeVoe) [reply planetterp/PlanetTerp.py]: Default values for these parameters is None, so if someone calls the method without passing a value we then pass it to urlencode which gives us a url like `https://api.planetterp.com/v1/professors?type=ta&limit=None`, which is invalid. So we can't pass keys with None values to urlencode.

I meant to replace the default values with better ones though (100 and 0) so this might not be necessary.

File: setup.py (hunk 10 above)
Comment (Liam-DeVoe) [reply setup.py]: This is a bit of an unfortunate hack to work around the following scenario:

* someone installs planetterp and does not have the requests module
* pip goes to install planetterp, which involves importing setup.py 
* `import setup` executes the import statements at the top of setup.py, which in turn execute our `import requests` line in `PlanetTerp.py`. This errors because the user doesn't have `requests` and the install fails
* the solution to this is either to have a `version.py` file and parse that manually to get the version (this is what I do in other projects) or define it in multiple places like we have here.

Further reading: https://stackoverflow.com/a/7071358

Comment (Liam-DeVoe) [reply setup.py]: To clarify, if we wanted it in one place we would put it in `__init__.py`, which means setup.py has to import it from init. This sets off the chain reaction of imports that ends in attempting, and failing, to import requests.

Comment (saewitz) [reply setup.py]: I don't fully understand the issue, but I don't like that we need to update the version in two places. I'm ok with a version.py file, at least for now.

Comment (Liam-DeVoe) [reply setup.py]: that's also my preferred resolution, will go ahead with that

**Review by ethan-schaffer — APPROVED** [review]

File: README.md [hunk 11]
```diff
@@ -1,10 +1,23 @@
...
 
-To use this code to access the API, simply download or clone this repository, and make your own file in the same folder as **PlanetTerp.py** and **sample_usage.py**.
+```python
+import planetterp
+
+course = planetterp.course(name="CMSC131", reviews=True)
+courses = planetterp.courses(department="CMSC", limit=2)
+prof = planetterp.professor(name="TBA", reviews="true")
+profs = planetterp.professors(type_="ta", limit=2)
+grades = planetterp.grades(course="CMSC131", professor="TBA")
```
Comment (Liam-DeVoe) [inline README.md]: you sure we want to use TBA? Might confuse people if they assume this is a special keyword with special meaning, instead of a normal professor (which probably shouldn't even be on the site)

Comment (saewitz) [reply README.md]: I'd prefer to use a not-real professor name because it isn't really fair to a professor to use them as an example. I agree it can be confusing -- maybe we can put a name there that isn't a professor instead.
